esiclient/tests/functional/base.py

```diff
@@ -0,0 +1,188 @@
+                      project_domain_name=self.projects[project]['domain'],
+                      identity_api_version='3',
+                      uri=self.config['auth_url']))
+
+    def _init_dummy_project(self, name, roles, parent=None):
+        admin_client = self.clients['admin']
+
+        # NOTE: names created using tempest's data_utils.rand_name() function
+        #       will be prefixed with 'tempest-' and suffixed with randomly
+        #       generated characters to prevent name collision.
+        project_name = data_utils.rand_name('esi-%s-project' % name)
+        flags = '--domain default --enable -f json'
+        if parent:
+            if parent not in self.projects.keys():
+                raise NameError('Invalid parent project: %s' % parent)
+            flags += ' --parent %s' % self.projects[parent]['name']
+        output = admin_client.openstack('project create %s' % project_name,
+                                        '', flags)
+        self._cls_cleanups.append([admin_client.openstack, 'project delete',
+                                   '', '%s' % project_name])
+        project_id = json.loads(output)['id']
+        self.projects[name] = {
+            'name': project_name,
+            'id': project_id,
+            'parent': parent,
+            'domain': 'default'
+        }
+
+        if roles is []:
+            raise ValueError('No roles specified when initializing dummy \
+                              project %s' % name)
+        elif type(roles) is str:
+            roles = [roles]
+
+        for role in roles:
+
+            username = data_utils.rand_name('esi-%s-%s' % (name, role))
+            password = data_utils.rand_password()
+            output = admin_client.openstack('user create %s' % username, '',
+                                            '--domain default --password %s \
+                                             --enable -f json' % password)
+            self._cls_cleanups.append([admin_client.openstack, 'user delete',
+                                       '', '%s' % username])
+            user_id = json.loads(output)['id']
+
+            client_name = '%s-%s' % (name, role)
+            self.users[client_name] = {
+                'name': username,
+                'id': user_id,
+                'password': password,
+                'domain': 'default'
+            }
+
+            admin_client.openstack('role add', '', '--user %s --project %s %s'
+                                   % (username, project_name, role))
+            self._cls_cleanups.append([admin_client.openstack, 'role remove',
+                                       '', '--user %s --project %s %s' %
+                                       (username, project_name, role)])
+
+            self._init_client(self, client_name, name)
+
+    def _verify_roles(self):
+        # Checks to see the is_node_owner and is_node_lessee roles exist
+        for role in 'owner', 'lessee':
```

Comment:
Should we add a check for roles named `esi_leap_owner` and `esi_leap_lessee`? I know that in the [esi-leap default policy](https://github.com/CCI-MOC/esi-leap/blob/master/esi_leap/common/policy.py) having these be the names for the is_node_owner and is_node_lessee roles is also a valid configuration.

Reply:
Well, as it stands, we are only testing ironic stuff here. The policy file I configured with my ironic instance is here: https://github.com/CCI-MOC/esi/blob/master/etc/ironic/policy.yaml.sample. This policy file actually doesnt do anything with roles, just checking that certain fields (owner and lessee field of a node) are equal to the requester's project id

Reply:
Ah, that makes sense! Do you plan on having initialization error out if said roles don't exist on the cloud being tested?

Reply:
dang. I guess I should gut this pr a bit then huh? Most of the work was in getting the dummy classes to work

Reply:
I think the code is useful in general; at the very least users should probably have the 'member' role as that's baseline, and when we eventually move to secure RBAC we may require additional roles.